File: DPath.java
/** Implementation of a depth first search
 */
public class DPath implements Path {
    private Map m;

    public DPath(Map m) {
        this.m = m;
    }

    public boolean pathExists(Point s, Point e) {
        m.clearMarks();
        return recPathExists(s, e);
    }
    private boolean recPathExists(Point s, Point e) {
        // Can't move here (impassable or already visited)
        if(!m.passible(s) || m.hasMark(s)) {
            return false;
        }
        // Note that we are visiting this location
        m.mark(s, '*');
        // Return true if we reached the end
        if(s.equals(e)) {
            return true;
        }
        // Try all the neighboring locations
        Point u = new Point(s.x, s.y-1);
        Point d = new Point(s.x, s.y+1);
        Point r = new Point(s.x+1, s.y);
        Point l = new Point(s.x-1, s.y);
        if(recPathExists(u, e)) {
            return true;
        }
        if(recPathExists(d, e)) {
            return true;
        }
        if(recPathExists(l, e)) {
            return true;
        }
        if(recPathExists(r, e)) {
            return true;
        }
        // Can't reach the end from any neighbors...
        return false;
    }
}

File: dpath.c
#include "path.h"
#include "map.h"
#include "point.h"
#include <stdio.h>
#include <stdlib.h>

struct Path* path_init(struct Map* map){
    struct Path* path = malloc(sizeof(struct Path));
    path->map = map;
    return path;
}
int recPathExists(struct Path* path, struct Point* s, struct Point* e){
    if ((passable(path->map, s) == 0) || (hasMark(path->map, s) == 1)){
        return 1;
    }
    mark(path->map, s, '0');
    if(equals(s, e) == 0) {
        return 0;
    }
    struct Point* u = point_init(s->x, s->y-1);
    struct Point* d = point_init(s->x, s->y+1);
    struct Point* r = point_init(s->x+1, s->y);
    struct Point* l = point_init(s->x-1, s->y);
    int result = recPathExists(path, u, e);
    free(u);
    if(result == 0) {
        return 0;
    }
    
    result = recPathExists(path, d, e);
    free(d);
    if(result == 0) {
        return 0;
    }
    result = recPathExists(path, l, e);
    free(l);
    if(result == 0) {
        return 0;
    }
    result = recPathExists(path, r, e);
    free(r);
    if(result == 0) {
        return 0;
    }
    return 1;
}
int pathExists(struct Path* path, struct Point* s, struct Point* e){
    clearMarks(path->map);
    return recPathExists(path, s, e);
}
void free_path(struct Path* path){
    free_map(path->map);
    free(path);
}
